### PR TITLE
[2125] Add the possibility to programmatically order and filter pages…

### LINKED ISSUE #2125: Add the possibility to programmatically order and filter the pages of a form

### PR DESCRIPTION
**Pull request by frouene**
… of a form.

Bug: https://github.com/eclipse-sirius/sirius-components/issues/2125

# Pull request template

## General purpose
What is the main goal of this pull request?
- [ ] Bug fixes
- [x] New features
- [ ] Documentation
- [ ] Cleanup
- [ ] Tests
- [ ] Build / releng

## Project management
- [ ] Has the pull request been added to the relevant project and milestone? (Only if you know that your work is part of a specific iteration such as the current one)
- [ ] Have the `priority:` and `pr:` labels been added to the pull request? (In case of doubt, start with the labels `priority: low` and `pr: to review later`)
- [ ] Have the relevant issues been added to the pull request?
- [ ] Have the relevant labels been added to the issues? (`area:`, `difficulty:`, `type:`)
- [ ] Have the relevant issues been added to the same project and milestone as the pull request?
- [ ] Has the `CHANGELOG.adoc` been updated to reference the relevant issues?
- [ ] Have the relevant API breaks been described in the `CHANGELOG.adoc`? (Including changes in the GraphQL API)
- [ ] In case of a change with a visual impact, are there any screenshots in the `CHANGELOG.adoc`? For example in `doc/screenshots/2022.5.0-my-new-feature.png`

## Architectural decision records (ADR)
- [ ] Does the title of the commit contributing the ADR start with `[doc]`?
- [ ] Are the ADRs mentioned in the relevant section of the  `CHANGELOG.adoc`?

## Dependencies
- [ ] Are the new / upgraded dependencies mentioned in the relevant section of the `CHANGELOG.adoc`?
- [ ] Are the new dependencies justified in the `CHANGELOG.adoc`?

## Frontend

This section is not relevant if your contribution does not come with changes to the frontend.

### General purpose
- [ ] Is the code properly tested? (Plain old JavaScript tests for business code and tests based on React Testing Library for the components)

### Typing
We need to improve the typing of our code, as such, we require every contribution to come with proper TypeScript typing for both changes contributing new files and those modifying existing files.
Please ensure that the following statements are true for each file created or modified (this may require you to improve code outside of your contribution).

- [ ] Variables have a proper type
- [ ] Functions’ arguments have a proper type
- [ ] Functions’ return type are specified
- Hooks are properly typed:
	- [ ] `useMutation<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useQuery<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useSubscription<DATA_TYPE, VARIABLE_TYPE>(…)`
	- [ ] `useMachine<CONTEXT_TYPE, EVENTS_TYPE>(…)`
	- [ ] `useState<STATE_TYPE>(…)`
- [ ] All components have a proper typing for their props
- [ ] No useless optional chaining with `?.` (if the GraphQL API specifies that a field cannot be `null`, do not treat it has potentially `null` for example)
- [ ] Nullable values have a proper type (for example `let diagram: Diagram | null = null;`)

## Backend

This section is not relevant if your contribution does not come with changes to the backend.

### General purpose
- [ ] Are all the event handlers tested?
- [ ] Are the event processor tested?
- [ ] Is the business code (services) tested?
- [ ] Are diagram layout changes tested?

### Architecture
- [ ] Are data structure classes properly separated from behavioral classes?
- [ ] Are all the relevant fields final?
- [ ] Is any data structure mutable? If so, please write a comment indicating why
- [ ] Are behavioral classes either stateless or side effect free?

### How to test this PR?
_Please describe here the various use cases to test this pull request_

- [ ] Has the Kiwi TCMS test suite been updated with tests for this contribution?

### CODE REVIEW

File: CHANGELOG.adoc [hunk 1]
```diff
@@ -30,8 +30,7 @@ This allows for code which does not depend on Sirius Web and the Document notion
 
 === Bug fixes
 
-- https://github.com/eclipse-sirius/sirius-components/issues/2058[#2058] [view] Fix an issue where the default icon for List widget candidates was missing when the candidates wer
-e not EObjects.
+- https://github.com/eclipse-sirius/sirius-components/issues/2058[#2058] [view] Fix an issue where the default icon for List widget candidates was missing when the candidates wer e not EObjects.
```
Comment (pcdavid) [inline CHANGELOG.adoc]: It fixes some formatting issue, but there is still an additional space in "wer e".

File: packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/api/IFormProcessor.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+
+import org.eclipse.sirius.components.forms.Form;
+import org.eclipse.sirius.components.representations.VariableManager;
+
+/**
+ * Interface to give the possibility to make changes on a form.
+ *
+ * @author frouene
+ */
+public interface IFormProcessor {
```
Comment (pcdavid) [inline packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/api/IFormProcessor.java]: Maybe `IFormPostProcessor#postProcess`? It's not clear from the current name when/on what it applies in the process.

File: packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        return this.formEventHandlers;
+    }
+
+    public IWidgetSubscriptionManagerFactory getWidgetSubscriptionManagerFactory() {
+        return this.widgetSubscriptionManagerFactory;
+    }
+
+    public Optional<IFormProcessor> getOptionalFormProcessor() {
+        return this.optionalFormProcessor;
+    }
```
Comment (pcdavid) [inline packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java]:
```java
public IFormProcessor getFormProcessor() {
        return this.optionalFormProcessor.orElse(new IFormProcessor.NoOp());
}
```
would avoid dealing with optionals and nullable fields in the rest of the code.

File: packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+@Configuration
+public class FormEventProcessorFactoryConfiguration {
+
+    private final IObjectService objectService;
+
+    private final List<IFormEventHandler> formEventHandlers;
+
+    private final IWidgetSubscriptionManagerFactory widgetSubscriptionManagerFactory;
+
+    private final Optional<IFormProcessor> optionalFormProcessor;
```
Comment (pcdavid) [inline packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java]: This means there can only be a single `IFormProcessor` globally in a given application, right?
Maybe it's fine, but it should be mentioned in the shape. If an application integrator needs to use different filtering logic to different kinds of forms, it's up to home to provide a single global `IFormProcessor` which detects the different cases and applies the proper treatment to each.

**Review by pcdavid — CHANGES_REQUESTED** [review]

File: packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java (hunk 4 above)
Comment (frouene) [reply packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java]: The alternative would be to add in the interface a `canProcess(Form form)` predicate to get the good IFormProcessor to use ?
What happen if multi IFormProcessor match, use the first one or for each apply process to the form...

Comment (pcdavid) [reply packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java]: I think it's OK to keep the API as is, just worth mentioning (in the CHANGELOG, in the doc, to potential users) the consequences: the application can only have a single IFormProcessor, if they need to apply different rules to different, it's up to them to provide a multiplexing/dispatch mechanism at the application level.

File: packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/api/IFormProcessor.java (hunk 2 above)
Comment (frouene) [reply packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/api/IFormProcessor.java]: I added the post to be clearer

File: packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java (hunk 3 above)
Comment (frouene) [reply packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java]: Done

File: packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java (hunk 4 above)
Comment (frouene) [reply packages/forms/backend/sirius-components-collaborative-forms/src/main/java/org/eclipse/sirius/components/collaborative/forms/configuration/FormEventProcessorFactoryConfiguration.java]: I've mentioned this constraint in the changelog

**Review by pcdavid — APPROVED** [review]

**Pull request merged**